File: deno/handlers/inbox.ts
import {
  Accept,
  Activity,
  Add,
  Announce,
  Block,
  Create,
  Delete,
  EmojiReact,
  Flag,
  Follow,
  Like,
  Move,
  Reject,
  Remove,
  Undo,
  Update,
  fetchDocumentLoader,
} from "@fedify/fedify";

export interface InboxPayload {
  raw: Record<string, unknown>;
}

export type InboxInstruction =
  | { action: "save"; object: unknown }
  | { action: "save_and_reply"; save: unknown; reply: unknown; inbox: string }
  | { action: "ignore" };

export async function handleInbox(payload: InboxPayload): Promise<InboxInstruction> {
  const documentLoader = fetchDocumentLoader;
  const raw = payload.raw;
  const type = raw["type"];

  if (type === "Follow") {
    const follow = await Follow.fromJsonLd(raw, { documentLoader });
    const actorId = follow.actorId;
    if (actorId == null) return { action: "ignore" };

    const followJson = await follow.toJsonLd({ contextLoader: documentLoader });
    return {
      action: "save_and_reply",
      save: { follow: followJson },
      reply: followJson,
      inbox: actorId.href,
    };
  }

  if (type === "Announce") {
    const announce = await Announce.fromJsonLd(raw, { documentLoader });
    const announceJson = await announce.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: announceJson };
  }
  if (type === "Create" || type === "Update" || type === "Delete") {
    let activity: Activity;
    if (type === "Create") {
      activity = await Create.fromJsonLd(raw, { documentLoader });
    } else if (type === "Update") {
      activity = await Update.fromJsonLd(raw, { documentLoader });
    } else {
      activity = await Delete.fromJsonLd(raw, { documentLoader });
    }
    const activityJson = await activity.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: activityJson };
  }

  if (type === "Like" || type === "EmojiReact") {
    const activity: Activity = type === "Like"
      ? await Like.fromJsonLd(raw, { documentLoader })
      : await EmojiReact.fromJsonLd(raw, { documentLoader });
    const activityJson = await activity.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: activityJson };
  }

  if (type === "Undo") {
    const undo = await Undo.fromJsonLd(raw, { documentLoader });
    const undoJson = await undo.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: undoJson };
  }

  if (type === "Accept" || type === "Reject") {
    const activity: Activity = type === "Accept"
      ? await Accept.fromJsonLd(raw, { documentLoader })
      : await Reject.fromJsonLd(raw, { documentLoader });
    const activityJson = await activity.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: activityJson };
  }

  if (type === "Move") {
    const move = await Move.fromJsonLd(raw, { documentLoader });
    const moveJson = await move.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: moveJson };
  }

  if (type === "Block") {
    const block = await Block.fromJsonLd(raw, { documentLoader });
    const blockJson = await block.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: blockJson };
  }

  if (type === "Flag") {
    const flag = await Flag.fromJsonLd(raw, { documentLoader });
    const flagJson = await flag.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: flagJson };
  }

  if (type === "Add" || type === "Remove") {
    const activity: Activity = type === "Add"
      ? await Add.fromJsonLd(raw, { documentLoader })
      : await Remove.fromJsonLd(raw, { documentLoader });
    const activityJson = await activity.toJsonLd({ contextLoader: documentLoader });
    return { action: "save", object: activityJson };
  }

  return { action: "ignore" };
}

File: elixir/lib/sukhi_fedi/fedi/inbox.ex
# SPDX-License-Identifier: AGPL-3.0-or-later
defmodule SukhiFedi.Fedi.Inbox do
  @moduledoc """
  Turns a verified inbound activity into a small typed instruction
  (`fedify.inbox.v1`) — the classify step of the inbox's
  verify → classify → execute pipeline (CODE_STYLE §2).

  Follow gets the special shape: it needs an `Accept(Follow)` reply and
  the follower's inbox, which means resolving the remote actor (signed
  when we have a local keypair, for Secure-Mode peers). Every other
  known type passes through as `save`; the executor
  (`AP.Instructions`) works on the activity JSON as delivered.

  The Bun version round-tripped activities through fedify's vocab
  classes here. For Mastodon-family senders that round trip is the
  identity transform (they already deliver compacted AS JSON), so the
  port hands the raw activity straight to the instruction and keeps
  classification to the one field it needs: `type`.
  """

  alias SukhiFedi.Fedi.Fetcher

  # Mirror of bun/fedify/activity_kinds.ts — Follow stays out of this
  # list because of its special reply shape.
  #
  # TODO(FEP-044f): hackers.pub and Hollo send `QuoteRequest` when one
  # of their users quotes a gated post, and expect a
  # `QuoteAuthorization` (or Reject) back — shaped like the Follow →
  # Accept flow below, so it belongs next to it. Until then their
  # quotes of our posts fall back to legacy handling.
  @generic_kinds ~w(Announce Create Update Delete Like EmojiReact Undo
                    Accept Reject Move Block Flag Add Remove)

  @type fetch_fun :: (String.t(), map() | nil -> {:ok, map()} | {:error, term()})

  @doc """
  Handles a `fedify.inbox.v1` payload and returns the instruction map
  (`action: save | save_and_reply | ignore`) the executor expects.
  """
  @spec handle(map(), fetch_fun()) :: {:ok, map()} | {:error, term()}
  def handle(payload, fetch_fun \\ &Fetcher.fetch_document/2)

  def handle(%{"raw" => raw} = payload, fetch_fun) when is_map(raw) do
    case raw["type"] do
      "Follow" -> follow_instruction(raw, payload, fetch_fun)
      kind when kind in @generic_kinds -> {:ok, %{"action" => "save", "object" => raw}}
      _ -> {:ok, %{"action" => "ignore"}}
    end
  end

  def handle(_payload, _fetch_fun), do: {:ok, %{"action" => "ignore"}}

  # The follower is waiting on the Accept — without it their
  # pending-follow state never resolves. Resolve their actor document
  # for the inbox URL, then hand back follow + reply + destination.
  defp follow_instruction(raw, payload, fetch_fun) do
    actor_uri = reference_uri(raw["actor"])
    followee_uri = reference_uri(raw["object"])

    with true <- is_binary(actor_uri) and is_binary(followee_uri),
         {:ok, %{"document" => actor}} <- fetch_fun.(actor_uri, payload["signAs"]),
         inbox when is_binary(inbox) <- actor["inbox"] do
      accept = %{
        "@context" => "https://www.w3.org/ns/activitystreams",
        "id" => accept_id(payload["selfDomain"] || URI.parse(followee_uri).host),
        "type" => "Accept",
        "actor" => followee_uri,
        "object" => %{
          "id" => raw["id"],
          "type" => "Follow",
          "actor" => actor_uri,
          "object" => followee_uri
        }
      }

      {:ok,
       %{
         "action" => "save_and_reply",
         "save" => %{"follow" => raw, "followeeUri" => followee_uri},
         "reply" => accept,
         "inbox" => inbox
       }}
    else
      # Missing ids or an unresolvable actor: same as the Bun handler,
      # ignore rather than fail the whole inbox POST.
      _ -> {:ok, %{"action" => "ignore"}}
    end
  end

  defp accept_id(domain) do
    "https://#{domain}/activities/accept/#{Ecto.UUID.generate()}"
  end

  # AP object references arrive as a bare IRI or an embedded object.
  defp reference_uri(uri) when is_binary(uri) and uri != "", do: uri
  defp reference_uri(%{"id" => uri}) when is_binary(uri) and uri != "", do: uri
  defp reference_uri(_), do: nil
end
